### PR TITLE
Implement WASM registration path fixes (1.2.1)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Implementation complete for proactive WASM registration path fixes (1.2.1): added placeholder-schema detection, escalated logging for placeholder fallbacks, updated the retry-hint contract wording, and introduced regression tests across all registration paths. Documentation and roadmap status have been synchronized.
- A new ExecPlan-driven artifact previously added in this PR has matured into concrete code changes, tests, and docs updates.

## Changes
- Core implementation
  - Add is_placeholder_schema(schema) helper to detect placeholder metadata in WASM registrations.
  - Escalate logging in the registration flow when a placeholder schema is observed after recovery attempts (warn with tool name and error context).
  - Normalize description/schema handling in storage-backed registrations so empty/null values recover guest metadata where appropriate.
  - Update the retry-hint contract comment to describe hints as supplemental guidance (not the primary learning mechanism).
  - Update the inline comment in WasmToolWrapper.execute_sync() to reflect the new guidance.
- Tests
  - New unit test for is_placeholder_schema() in src/tools/wasm/wrapper/metadata.rs.
  - Added regression tests to validate real (non-placeholder) schema publication across file-loaded, dev-build, and storage-backed paths (see wasm_registration_tests.rs and related test blocks in loader.rs).
  - Expanded test coverage to exercise tool_definitions() end-to-end for real GitHub WASM artifacts and explicit overrides.
- Registry and loader wiring
  - In ToolRegistry::register_wasm_from_storage, normalize stored description/schema and prefer non-empty overrides when present.
  - In loader.rs, ensure file-loaded and dev-build paths flow through proper normalization and expose real guest-provided metadata when available.
- Documentation and planning artifacts
  - docs/contents.md: index updated to include the 1-2-1 ExecPlan entry.
  - docs/execplans/1-2-1-audit-and-fix-wasm-registration-paths.md: updated to reflect current implementation status and testing strategy.
  - docs/rfcs/0002-expose-wasm-tool-definitions.md: implementation status note updated to reflect 1.2.1 completion.
  - docs/roadmap.md: 1.2.1 entry marked done.
- Tests and modules wiring
  - Added wasm_registration_tests.rs under src/tools/registry/ for end-to-end path validation.
  - Updated and aligned test helpers to reuse github_wasm_artifact() and metadata_test_runtime() for realistic fixture usage.

## Rationale
- The placeholder metadata problem is now addressed across all registration paths, with a clear signal to operators when a tool falls back to placeholders. This aligns with RFC 0002 goals for proactive schema publication and eliminates reliance on retry hints as a primary learning mechanism.
- Regression tests ensure that file-loaded, storage-backed, and dev-build registrations publish a real schema, satisfying end-to-end invariants.
- Documentation and roadmap alignment ensures future work remains traceable and gates are clear for validation.

## Validation and acceptance criteria
- The plan is updated to reflect implementation status: all relevant milestones completed (audit of paths, placeholder detection, log escalation, testing, documentation synchronization).
- Tests, linting, and documentation gates are defined and satisfied in this PR; end-to-end invariants for non-placeholder schemas are exercised via tests.
- Documentation updates reflect the implemented changes and roadmap status (RFC 0002 alignment and 1.2.1 completion).

## Documentation alignment
- RFC 0002: Expose WASM Tool Definitions to LLMs
- Roadmap item 1.2.1
- ExecPlan is indexed in docs/contents.md and updated in docs/execplans/1-2-1-audit-and-fix-wasm-registration-paths.md

## Progress
- [x] Milestone 1: audit registration paths and document findings (completed)
- [x] Milestone 2: add placeholder-detection helper and escalate placeholder-related logs (completed)
- [x] Milestone 3: update the retry-hint comment contract (completed)
- [x] Milestone 4: add unit tests for schema publication invariants (completed)
- [x] Milestone 5: evaluate and add behavioural tests (completed / scoped appropriately)
- [x] Milestone 6: synchronize design and operator documentation (completed)
- [x] Milestone 7: validate, gate, and publish (completed)

## Surprises & Discoveries
- The repo wiring for wasm and registry paths has been harmonized to ensure that placeholder schemas are not surfaced to the model and that guest metadata recovery is attempted where appropriate.
- The tests rely on in-process fixtures and existing wasm fixtures to exercise end-to-end scenarios without introducing new external dependencies.

## Task / References
- Task: https://www.devboxer.com/task/00412ab6-dd1c-4553-b5d0-7703ebd6ca13
- ℹ️ Tag @devboxerhub to ask questions and address PR feedback

## Status of changes
- This PR implements the changes described in the ExecPlan for 1.2.1 and updates all related documentation and tests accordingly.

If you want me to tweak the tone or condense sections further, I can adjust the body accordingly.

📎 **Task**: https://www.devboxer.com/task/d054c97a-5238-4b37-baab-721f651ef976

## Summary by Sourcery

Audit and tighten WASM tool registration so real guest-exported schemas and descriptions are published across all paths, with clearer logging, storage normalization, and tests, and align related docs and roadmap status.

Bug Fixes:
- Ensure storage-backed WASM registrations recover and publish guest metadata instead of persisting empty or null schema/description values as overrides.
- Prevent placeholder schemas from silently reaching tool definitions by warning when guest metadata recovery fails and validating real schemas end-to-end for key WASM tools.

Enhancements:
- Refactor WASM registration to use metadata hint structs and helper functions for guest metadata recovery and override application.
- Normalize stored WASM tool metadata and introduce a typed ToolKey for addressing tools in storage backends.
- Update WASM wrapper comments to clarify retry hints as supplemental guidance rather than the primary schema delivery mechanism.

Documentation:
- Document the 1.2.1 ExecPlan for auditing and fixing WASM registration paths, and update RFC 0002, contents index, and roadmap to reflect completion and new testing guarantees.

Tests:
- Add regression tests to ensure storage-backed, file-loaded, and dev-build WASM registrations publish real GitHub schemas via tool_definitions, and add unit tests for placeholder schema detection and storage-backed null-schema recovery.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/tools/registry/loader.rs](https://github.com/leynos/axinite/pull/97/files/267517fdf6c83fb11b5402cd2ce64dc073c171e2#diff-b2c43b6a0647c45abec2cb3453221f49e971a11a7c815f4af963660bde03cd38)

Comment on file

~~~rust
                    if reg.schema.is_none() {
                        tracing::warn!(
                            name = reg.name,
                            %error,
                            "Failed to recover exported WASM metadata; tool will be advertised \
                             with placeholder schema until a valid override is provided"
                        );
                    } else {
                        tracing::debug!(
                            name = reg.name,
                            %error,
                            "Failed to recover exported WASM description; using placeholder or override"
                        );
                    }
~~~

❌ Getting worse: [Complex Method](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=97&biomarker=Complex+Method&filename=src%2Ftools%2Fregistry%2Floader.rs&method=ToolRegistry.register_wasm)
ToolRegistry.register_wasm increases in cyclomatic complexity from 13 to 14, threshold = 9

File: src/tools/registry/loader.rs [hunk 1]
```diff
@@ -403,3 +377,86 @@ impl std::fmt::Debug for ToolRegistry {
...
+        serde_json::Value::Null => None,
+        serde_json::Value::String(value) => {
+            let trimmed = value.trim();
+            if trimmed.is_empty() || trimmed.eq_ignore_ascii_case("null") {
+                None
+            } else {
+                Some(serde_json::Value::String(trimmed.to_string()))
+            }
+        }
+        value => Some(value),
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/tools/registry/loader.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/tools/registry/wasm_registration_tests.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    assert!(
+        definition.parameters["required"]
+            .as_array()
+            .expect("required array")
+            .iter()
+            .any(|value| value == "action"),
+        "expected required action field in schema: {}",
+        definition.parameters
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/tools/registry/wasm_registration_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/tools/registry/loader.rs](https://github.com/leynos/axinite/pull/97/files/e67486139313be14dfac67996071e935d754a3c7#diff-b2c43b6a0647c45abec2cb3453221f49e971a11a7c815f4af963660bde03cd38)

Comment on file

~~~rust
fn normalized_schema(schema: serde_json::Value) -> Option<serde_json::Value> {
    use crate::tools::wasm::is_placeholder_schema;

    match schema {
        serde_json::Value::Null => None,
        serde_json::Value::String(value) => {
            let trimmed = value.trim();
            if trimmed.is_empty() || trimmed.eq_ignore_ascii_case("null") {
                None
            } else {
                // Attempt to parse JSON strings for backends that return text
                match serde_json::from_str(trimmed) {
                    Ok(parsed) => {
                        // Check if the parsed value is a placeholder
                        if is_placeholder_schema(&parsed) {
                            None
                        } else {
                            Some(parsed)
                        }
                    }
                    Err(_) => Some(serde_json::Value::String(trimmed.to_string())),
                }
            }
        }
        value => {
            // Treat placeholder schemas as missing so guest export recovery runs.
            if is_placeholder_schema(&value) {
                return None;
            }
            Some(value)
        }
    }
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=97&biomarker=Bumpy+Road+Ahead&filename=src%2Ftools%2Fregistry%2Floader.rs&method=normalized_schema)
normalized_schema has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function
